Configuration property name 'spring.shardingsphere.rules.sharding.sharding-algorithms.XXX.props' is not valid

**Question**
Caused by: org.springframework.boot.context.properties.source.InvalidConfigurationPropertyNameException: Configuration property name 'spring.shardingsphere.rules.sharding.sharding-algorithms.user_record_standard.props' is not valid

**Version**
Springboot2.2.9 | mybatis-plus| shardingsphere-jdbc-core-spring-boot-starter 5.0.0-alpha

**Config**
```
spring:
    shardingsphere:
    rules:
      sharding:
        default-data-source-name: ds0
        tables:
          user_record:
            actual-data-nodes: ds0.user_record,ds0.user_record_$->{2019..2022}
            table-strategy:
              standard:
                sharding-column: user_id
                sharding-algorithm-name: user_record_standard
        sharding-algorithms:
          user_record_standard:
            type: UserRecordShardingAlgorithm
            props:
              allow-range-query-with-record-sharding: false

```

**Exception**

```
Exception in thread "main" java.lang.reflect.UndeclaredThrowableException
	at org.springframework.util.ReflectionUtils.rethrowRuntimeException(ReflectionUtils.java:147)
	at org.springframework.boot.SpringApplication.handleRunFailure(SpringApplication.java:810)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:325)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215)
	at cn.wx.provider.SmartlearningApplication.main(SmartlearningApplication.java:19)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.shardingsphere.spring.boot.util.PropertyUtil.v2(PropertyUtil.java:109)
	at org.apache.shardingsphere.spring.boot.util.PropertyUtil.handle(PropertyUtil.java:75)
	at org.apache.shardingsphere.spring.boot.registry.AbstractAlgorithmProvidedBeanRegistry.lambda$registerBean$1(AbstractAlgorithmProvidedBeanRegistry.java:57)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at org.apache.shardingsphere.spring.boot.registry.AbstractAlgorithmProvidedBeanRegistry.registerBean(AbstractAlgorithmProvidedBeanRegistry.java:55)
	at org.apache.shardingsphere.sharding.spring.boot.algorithm.ShardingAlgorithmProvidedBeanRegistry.postProcessBeanDefinitionRegistry(ShardingAlgorithmProvidedBeanRegistry.java:38)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:280)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:126)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:707)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:533)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:141)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
	... 3 more
Caused by: org.springframework.boot.context.properties.source.InvalidConfigurationPropertyNameException: Configuration property name 'spring.shardingsphere.rules.sharding.sharding-algorithms.user_record_standard.props' is not valid
	at org.springframework.boot.context.properties.source.ConfigurationPropertyName.elementsOf(ConfigurationPropertyName.java:522)
	at org.springframework.boot.context.properties.source.ConfigurationPropertyName.elementsOf(ConfigurationPropertyName.java:499)
	at org.springframework.boot.context.properties.source.ConfigurationPropertyName.of(ConfigurationPropertyName.java:490)
	at org.springframework.boot.context.properties.source.ConfigurationPropertyName.of(ConfigurationPropertyName.java:478)
	at org.springframework.boot.context.properties.bind.Binder.bind(Binder.java:199)
	at org.springframework.boot.context.properties.bind.Binder.bind(Binder.java:186)
	... 21 more
```

@terrymanu 

**Sorry, because of my log settings, the log is not printed completely, and the following content is missing**
![image](https://user-images.githubusercontent.com/26064998/100056392-9666f500-2e60-11eb-8c6d-cf4e93499f8b.png)

**So, when I changed user_record_standard to user-record-standard, the error no longer reappeared.
But another error occurred**

```
Exception in thread "main" java.lang.reflect.UndeclaredThrowableException
	at org.springframework.util.ReflectionUtils.rethrowRuntimeException(ReflectionUtils.java:147)
	at org.springframework.boot.SpringApplication.handleRunFailure(SpringApplication.java:810)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:325)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215)
	at cn.wx.provider.SmartlearningApplication.main(SmartlearningApplication.java:19)

Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.shardingsphere.spring.boot.util.PropertyUtil.v2(PropertyUtil.java:111)
	at org.apache.shardingsphere.spring.boot.util.PropertyUtil.handle(PropertyUtil.java:75)
	at org.apache.shardingsphere.spring.boot.registry.AbstractAlgorithmProvidedBeanRegistry.registerBean(AbstractAlgorithmProvidedBeanRegistry.java:50)
	at org.apache.shardingsphere.sharding.spring.boot.algorithm.KeyGenerateAlgorithmProvidedBeanRegistry.postProcessBeanDefinitionRegistry(KeyGenerateAlgorithmProvidedBeanRegistry.java:38)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:280)
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:126)
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:707)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:533)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:141)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315)
	... 3 more

Caused by: java.util.NoSuchElementException: No value bound
	at org.springframework.boot.context.properties.bind.BindResult.get(BindResult.java:56)
	... 19 more
```


**I saw other same issues, and changed the config as suggested**

![image](https://user-images.githubusercontent.com/26064998/100057972-6ec55c00-2e63-11eb-8a86-82b75e97a0c4.png)

**But did not solve the problem**

--- a/shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-spring-infra/shardingsphere-jdbc-spring-boot-starter-infra/src/main/java/org/apache/shardingsphere/spring/boot/registry/AbstractAlgorithmProvidedBeanRegistry.java
+++ b/shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-spring-infra/shardingsphere-jdbc-spring-boot-starter-infra/src/main/java/org/apache/shardingsphere/spring/boot/registry/AbstractAlgorithmProvidedBeanRegistry.java
@@ -43,41 +43,46 @@ import java.util.stream.Collectors;
 @RequiredArgsConstructor(access = AccessLevel.PROTECTED)
 public abstract class AbstractAlgorithmProvidedBeanRegistry<T extends ShardingSphereAlgorithm> implements BeanDefinitionRegistryPostProcessor, BeanPostProcessor {
     
-    private static final String TYPE_SUFFIX = ".type";
-    
-    private static final String PROPS_SUFFIX = ".props";
+    private static final String POINT = ".";
     
     private static final String PROPS = "props";
+    
+    private static final String PROPS_SUFFIX = POINT + PROPS;
+    
+    private static final String TYPE_SUFFIX = ".type";
     
     private final Environment environment;
     
     @SuppressWarnings("all")
     protected void registerBean(final String prefix, final Class<T> algorithmClass, final BeanDefinitionRegistry registry) {
-        Map<String, Object> paramMap = PropertyUtil.handle(environment, prefix, Map.class);
-        Set<String> keys = paramMap.keySet().stream().map(key -> {
-            return key.contains(".") ? key.substring(0, key.indexOf(".")) : key;
-        }).collect(Collectors.toSet());
-        Map<String, YamlShardingSphereAlgorithmConfiguration> shardingAlgorithmMap = new LinkedHashMap<>();
-        keys.forEach(each -> {
-            YamlShardingSphereAlgorithmConfiguration config = new YamlShardingSphereAlgorithmConfiguration();
-            String propsPrefix = String.join("", prefix, each, PROPS_SUFFIX);
-            boolean exist = PropertyUtil.containPropertyPrefix(environment, propsPrefix);
-            if (exist) {
-                Map<String, Object> propsMap = PropertyUtil.handle(environment, propsPrefix, Map.class);
-                config.getProps().putAll(propsMap);
-            }
-            String typePrefix = String.join("", prefix, each, TYPE_SUFFIX);
-            String algorithmType = environment.getProperty(typePrefix);
-            config.setType(algorithmType);
-            shardingAlgorithmMap.put(each, config);
-        });
-        ShardingSphereServiceLoader.register(algorithmClass);
-        shardingAlgorithmMap.forEach((key, algorithmConfiguration) -> {
-            ShardingSphereAlgorithm algorithm = TypedSPIRegistry.getRegisteredService(algorithmClass, algorithmConfiguration.getType(), algorithmConfiguration.getProps());
-            BeanDefinitionBuilder builder = BeanDefinitionBuilder.genericBeanDefinition(algorithm.getClass());
-            builder.addPropertyValue(PROPS, algorithmConfiguration.getProps());
-            registry.registerBeanDefinition(key, builder.getBeanDefinition());
-        });
+        boolean existPrefix = PropertyUtil.containPropertyPrefix(environment, prefix);
+        if (existPrefix) {
+            Map<String, Object> paramMap = PropertyUtil.handle(environment, prefix, Map.class);
+            Set<String> keys = paramMap.keySet().stream().map(key -> {
+                return key.contains(POINT) ? key.substring(0, key.indexOf(POINT)) : key;
+            }).collect(Collectors.toSet());
+            Map<String, YamlShardingSphereAlgorithmConfiguration> shardingAlgorithmMap = new LinkedHashMap<>();
+            keys.forEach(each -> {
+                YamlShardingSphereAlgorithmConfiguration config = new YamlShardingSphereAlgorithmConfiguration();
+                String propsPrefix = String.join("", prefix, each, PROPS_SUFFIX);
+                boolean existProps = PropertyUtil.containPropertyPrefix(environment, propsPrefix);
+                if (existProps) {
+                    Map<String, Object> propsMap = PropertyUtil.handle(environment, propsPrefix, Map.class);
+                    config.getProps().putAll(propsMap);
+                }
+                String typePrefix = String.join("", prefix, each, TYPE_SUFFIX);
+                String algorithmType = environment.getProperty(typePrefix);
+                config.setType(algorithmType);
+                shardingAlgorithmMap.put(each, config);
+            });
+            ShardingSphereServiceLoader.register(algorithmClass);
+            shardingAlgorithmMap.forEach((key, algorithmConfiguration) -> {
+                ShardingSphereAlgorithm algorithm = TypedSPIRegistry.getRegisteredService(algorithmClass, algorithmConfiguration.getType(), algorithmConfiguration.getProps());
+                BeanDefinitionBuilder builder = BeanDefinitionBuilder.genericBeanDefinition(algorithm.getClass());
+                builder.addPropertyValue(PROPS, algorithmConfiguration.getProps());
+                registry.registerBeanDefinition(key, builder.getBeanDefinition());
+            });
+        }
     }
     
     @Override

